AppTaskQueue/appscale/taskqueue/service_manager.py

```diff
@@ -0,0 +1,154 @@
+""" Keeps track of queue configuration details for producer connections. """
+
+import json
+
+from tornado.ioloop import IOLoop
+
+from .utils import logger
+
+
+class ProjectServiceManager(dict):
+  """ Keeps track of service configuration details for a single project. """
+  def __init__(self, zk_client, db_access, project_id):
+    """ Creates a new ProjectServiceManager.
+
+    Args:
+      zk_client: A KazooClient.
+      db_access: A DatastoreProxy.
+      project_id: A string specifying a project ID.
+    """
+    super(ProjectServiceManager, self).__init__()
+    self.zk_client = zk_client
+    self.db_access = db_access
+    self.project_id = project_id
+    services_node = '/appscale/projects/{}/services'.format(self.project_id)
+    zk_client.ensure_path(services_node)
+    self.watch = zk_client.ChildrenWatch(services_node,
+                                         self._update_services_watch)
+
+  def update_services(self, new_services_list):
+    """ Establishes watches for all existing services.
+
+    Args:
+      new_services_list: A list of strings specifying existing service IDs.
+    """
+    to_remove = [service for service in self if service not in
+                 new_services_list]
+    for service_id in to_remove:
+      logger.debug("updating: {}".format(to_remove))
+      self[service_id].stop()
```

Comment:
In order to prevent a watch from resetting after it fires, return `False` from the registered function. Here's an example: https://github.com/AppScale/appscale/pull/2520/files#diff-5cdbc79a945a409d5709a087a1270767R141